### PR TITLE
Flexible tree type structure

### PR DESCRIPTION
**Pull request by AlexeyPerevalov**
This PR was created just for discussing.

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

### CODE REVIEW

File: pkg/apis/topology/v1alpha1/types.go [hunk 1]
```diff
@@ -29,24 +29,32 @@ type NodeResourceTopology struct {
...
+type Zone struct {
+	Type       string
+	Parent     string
+	Costs      map[int]int
+	Attributes map[string]int
+	Resources  ResourceInfoMap
 }
 
+type ResourceInfo struct {
+	Allocatable int
```
Comment (swatisehgal) [inline pkg/apis/topology/v1alpha1/types.go]: Shouldn't allocatable and capacity be of the type v1.ResourceList ?

Comment (swatisehgal) [reply pkg/apis/topology/v1alpha1/types.go]: At the very least, I think string would be more appropriate than integer given that we might have values such as 64Mi, 500m ?

File: pkg/apis/topology/v1alpha1/types.go [hunk 2]
```diff
@@ -29,24 +29,32 @@ type NodeResourceTopology struct {
...
+	Zones            map[string]Zone `json:"zones"`
 }
 
-// NUMANodeResource is the spec for a NodeResourceTopology resource
-type NUMANodeResource struct {
-	NUMAID int
-	Resources v1.ResourceList
+// Zone is the spec for a NodeResourceTopology resource
+type Zone struct {
+	Type       string
```
Comment (swatisehgal) [inline pkg/apis/topology/v1alpha1/types.go]: No json tags here for keys?
```
type Zone struct {
 Type string  `json:"type"`
 Parent string  `json:"parent"`
 Cost map[string]int  `json:"cost"`
 Attributes map[string]string  `json:"attributes"
 Resources  ResourceInfoMap  `json:"resources"
}
```

Comment (AlexeyPerevalov) [reply pkg/apis/topology/v1alpha1/types.go]: As I remember if json is omitted, json field name  is given as golang field name, but we need to specify is that field mandatory or not, so yes. I'll add json. Thanks.

File: pkg/apis/topology/v1alpha1/types.go (hunk 1 above)
Comment (AlexeyPerevalov) [reply pkg/apis/topology/v1alpha1/types.go]: yes, it's more practical.

**Pull request merged**